packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2737,6 +2737,8 @@ export class ContainerRuntime
 				return { continue: true };
 			};
 
+			this._summarizer?.markSummaryAttempt?.(summaryRefSeqNum);
```

Comment:
I am not sure I understand why this is required. The summary result has the correct reference sequence number so that can be used to correctly record an attempt.
My concern with this is that it is adding a dependency which is not very natural - i.e., calling back the caller in the middle of the operation it is doing. And not doing this correctly will result in summaries being totally wrong.

Reply:
This used to work fine before the changes to how heuristics track ops. I believe that is where this needs to be fixed.

Reply:
This also means that the caller is totally dependent on the calling code in order to function correctly, which increases the surface area of where things can go easily wrong.

Reply:
The issue is that the `submitSummary` method is resuming the inbound queue before marking the summary. This causes the heuristics to be incorrect.

For example:
1. summary ref seq num is 3
2. summary is submitted and returned from `submitSummary` call in the `SummaryGenerator`
3. inbound queues are resumed
4. 10 more runtime ops are processed by heuristics
5. `SummaryGenerator` records the summary attempt and considers those extra 10 ops as also being part of the summary

I can try moving the call to `summarizer.recordSummaryAttempt` right before we resume the inbound queues, but it needs to happen before resuming to ensure we are recording heuristics properly.

Reply:
I understand the problem but adding this indirect dependency is dangerous and can easily break summaries.

Reply:
I do not fully understand how the SummarizerHeuristics tracks ops now. But here's one something you could probably do:
- Queue the seq#s of ops received in SummarizerHeuristics.
- When record attempt is called, pass in the referenceSeq# for the completed summary. As part of this you can filter out the ops before and after this referenceSeq#. The ones before are part of this summary and the ones after are part of the next summary.

Reply:
If keeping the current logic of container runtime calling back into the summarizer, it should be encoded somehow as a dependency on the submitSummary function. And not doing it should result in a fail fast.

Reply:
I agree that the solution you pointed out could be safer in the long run. Should be noted that size complexity will increase and a couple more methods need to be exposed on heuristic data class (and will be doing some of these calculations).

I'll try giving it a shot, but if it proves to be complex I think we should commit this change as-is and revisit.

@NicholasCouri FYI

Reply:
I agree with you - it is not worthy it to make it more complicated than it already is.